File: week2/DeleteNodeLinkedList/deleteNodeLinkedList.cpp
#include <iostream>
using namespace std;

struct ListNode{
    int value;
    ListNode* next;
    ListNode():value(0), next(nullptr){}
    ListNode(int x):value(x), next(nullptr){}
    ListNode(int x, ListNode* next):value(x), next(next){}
};

ListNode* deleteNode(ListNode* input, int taregt);

// head = [4,5,1,9], node = 5 Output: [4,1,9]
int main(){
    int taregt = 5;
    ListNode* input1=new ListNode(9, nullptr);
    ListNode* input2=new ListNode(1, nullptr);
    ListNode* input3=new ListNode(5, nullptr);
    ListNode* input4=new ListNode(4, nullptr);
    input1->next=input2;
    input2->next=input3;
    input3->next=input4;
    input4->next=NULL;

    ListNode* output = deleteNode(input1, taregt);
    while(output){
        cout << output->value << "\n";
        output=output->next;
    }
}

ListNode* deleteNode(ListNode* input, int taregt){
    ListNode* output = new ListNode();
    output = input;
    ListNode* dummy = output;
    while(dummy){
        cout << dummy->value << "\n";
        if(dummy->value == taregt){
            dummy->value = dummy->next->value;
            dummy->next = dummy->next->next;
        }
        dummy=dummy->next;
    }
    return output;
}

File: week2/DeleteNodeLinkedList/deleteNodeLinkedList.py
from typing import Optional
class ListNode:
    @classmethod
    def __intit__(self, value, next):
        self.value = value
        self.next = next
class Solution:
    @classmethod
    def deleteNode(self, input: Optional[ListNode], target:int)->Optional[ListNode]:
        output = input
        dummy = output
        while dummy:
            if dummy.value == target:
                dummy.value = dummy.next.value
                dummy.next = dummy.next.next

            dummy=dummy.next
        return output
    
#head = [4,5,1,9], node = 5 Output: [4,1,9]
input1 = ListNode()
input1.value=9
input1_1 = ListNode()
input1_1.value=1
input1_2 = ListNode()
input1_2.value=5
input1_3 = ListNode()
input1_3.value=4
input1.next=input1_1
input1_1.next=input1_2
input1_2.next=input1_3
input1_3.next=None
target=5
output = ListNode()
output = Solution.deleteNode(input1, target)

while output:
    print(output.value)
    output=output.next